File: bestia-client/src/Bnet/Message/Crafting/CraftItemCMSG.cs
using Bnet;

namespace BestiaBehemothClient.Bnet.Message.Crafting
{
  /// <summary>
  /// Start one craft. Sent from the crafting window, which opened because the player activated a
  /// crafting skill - activation and execution are separate messages because
  /// <c>ActivateSkillCMSG</c> carries no recipe.
  ///
  /// The server may refuse (unknown recipe, skill too low, no station in range, materials short,
  /// already crafting) with an <c>OperationError</c>. Nothing is applied locally either way.
  /// </summary>
  public partial class CraftItemCMSG : ICMSG
  {
    public ulong RecipeId { get; set; }

    /// <summary>
    /// The item instance to work on, or 0 for a recipe that makes something instead.
    /// </summary>
    public ulong TargetUniqueId { get; set; }

    public override Envelope ToEnvelope()
    {
      var craftItem = new global::Bnet.CraftItemCMSG
      {
        RecipeId = RecipeId,
        TargetUniqueId = TargetUniqueId
      };

      return new Envelope
      {
        CraftItem = craftItem
      };
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/crafting/CraftItemCMSG.kt
package net.bestia.zone.crafting

import net.bestia.bnet.proto.CraftItemCmsgProto
import net.bestia.zone.message.CMSG

/**
 * The player picked a recipe in the crafting window and wants it made.
 *
 * Separate from `ActivateSkillCMSG` because that message carries no recipe: a skill that unlocks nine recipes
 * cannot say which one is wanted, so activation opens the window (see [CraftableRecipesSMSG]) and this executes
 * one line of it.
 */
data class CraftItemCMSG(
  override val playerId: Long,
  val recipeId: Long,

  /** The item instance to work on, or 0 for a recipe that makes something instead. */
  val targetUniqueId: Long
) : CMSG {
  companion object {
    fun fromBnet(accountId: Long, craftItem: CraftItemCmsgProto.CraftItemCMSG): CraftItemCMSG =
      CraftItemCMSG(
        playerId = accountId,
        recipeId = craftItem.recipeId,
        targetUniqueId = craftItem.targetUniqueId
      )
  }
}
